src/xinput.cpp

```diff
@@ -0,0 +1,126 @@
+    {
+        auto& interpreter = xeus::get_interpreter();
+
+        // Register the input handler
+        std::string value;
+        interpreter.register_input_handler([&value](const std::string& v) {
+            value = v;
+        });
+
+        // Send the input request.
+        interpreter.input_request(prompt, b);
+
+        // Remove input handler;
+        interpreter.register_input_handler(nullptr);
+
+        // TODO: Handle EOF
+        // if value == '\x04':
+        //    raise EOFError
+
+        return value;
+    }
+
+    // Free functions for intput, raw_input and getpass
+    std::string input(const std::string& prompt)
+    {
+        return input_request(prompt, false);
+    }
+
+    std::string getpass(const std::string& prompt)
+    {
+        return input_request(prompt, true);
+    }
+
+    void notimplemented(const std::string&)
+    {
+        throw std::runtime_error("This frontend does not support input requests");
+    }
+
+    PYBIND11_EMBEDDED_MODULE(xeus_python_input, m)
+    {
+        m.def("input", input, py::arg("prompt") = "")
+         .def("getpass", getpass, py::arg("prompt") = "")
+         .def("notimplemented", notimplemented, py::arg("prompt") = "");
+
+        if (PY_MAJOR_VERSION == 2)
+        {
+            m.def("raw_input", input, py::arg("prompt") = "");
+        }
+    }
+
+    // Implementation of input_redirection
+
+#if PY_MAJOR_VERSION == 2
+#define XEUS_PYTHON_BUILTINS __builtin__
+#else
+#define XEUS_PYTHON_BUILTINS builtins
+#endif
+
+    input_redirection::input_redirection(bool allow_stdin)
+    {
+        py::module xeus_python_input = py::module::import("xeus_python_input");
+
+        // Forward input()
+        py::module builtins = py::module::import(XPYT_STRINGIFY(XEUS_PYTHON_BUILTINS));
+        m_sys_input = builtins.attr("input");
+        builtins.attr("input") = allow_stdin ? xeus_python_input.attr("input")
+                                             : xeus_python_input.attr("notimplemented");
+
+        // Forward raw_input()
+        if (PY_MAJOR_VERSION == 2)
+        {
+            m_sys_raw_input = builtins.attr("raw_input");
+            builtins.attr("raw_input") = allow_stdin ? xeus_python_input.attr("raw_input")
+                                                     : xeus_python_input.attr("notimplemented");
+        }
+
+        // Forward getpass()
+        py::module getpass = py::module::import("getpass");
+        m_sys_getpass = getpass.attr("getpass");
+        getpass.attr("getpass") = allow_stdin ? xeus_python_input.attr("getpass")
+                                              : xeus_python_input.attr("notimplemented");
+    }
+
+    input_redirection::~input_redirection()
+    {
+        // Restore input()
+        py::module builtins = py::module::import(XPYT_STRINGIFY(XEUS_PYTHON_BUILTINS));
+        builtins.attr("input") = m_sys_input;
+
+        // Restore raw_input()
+        if (PY_MAJOR_VERSION == 2)
```

Comment:
`#if PY_MAJOR_VERSION <= 2`?

Reply:
I don't think we will ever support Python 1!

Reply:
I'm not talking about Python1. What I meant is that we could check that at compilation time

Reply:
Yes, that is what I was doing earlier. Why not!